lib/package-managers/npm.js

```diff
@@ -9,21 +10,46 @@ const packageJson  = require('package-json');
 const ProxyAgent   = require('https-proxy-agent');
 const npmConf      = require('npm-conf')();
 const proxyUrl     = require('get-proxy')();
+const registryUrl  = require('registry-url');
 
-let agent;
+const mainRegistryUrl = registryUrl();
+
+let proxyAgent;
+let noProxyEnv;
+let noProxyRegExps;
 if (proxyUrl) {
-    const proxy = new URL(proxyUrl);
-    agent = new ProxyAgent({
+    const proxy = global.URL ? new URL(proxyUrl) : url.parse(proxyUrl);
+    proxyAgent = new ProxyAgent({
         host: proxy.hostname,
         port: proxy.port,
         secureProxy: proxy.protocol === 'https:',
         keepAlive: true,
-        maxSockets: npmConf.get('maxsockets') || 50, // https://docs.npmjs.com/misc/config#maxsockets
+        maxSockets: npmConf.get('maxsockets'),
         ca: npmConf.get('ca'),
         key: npmConf.get('key'),
         cert: npmConf.get('cert'),
         rejectUnauthorized: npmConf.get('strict-ssl')
     });
+
+    noProxyEnv = process.env.NO_PROXY || process.env.no_PROXY || process.env.no_proxy;
```

Comment:
This block is still missing a high-level of explanation of *why* this code exists. Low-level comments are fine when they clarify a confusing statement, but not when they are a literal translation of the code (`Replace dots with escaped dots and asterisks with .* matchers `). Someone who is not familiar with proxies would still have no idea *why* this code is here. We need a short description of what the code does generally, and how it is applied specifically in ncu.

For example:

> The NO_PROXY environment variable specifies URLs that should be excluded from proxying. Generate regular expressions for each NO_PROXY url to be tested against the registry URL later when fetching package info.

Reply:
You're right. Thanks for providing a useful comment for that chunk of code, I've added it accordingly and also linked the no_proxy documentation from curl for more details.